# **NLP** FISAC-2---NLTK Taggers.


### Name-Utkarsh Aggarwal
### Registration number-200968234
### Section-'A'

----------------------------------------------------------------------------------------------------------------------------------------------------------------

#### Assignment:-

Create a regular expression tagger and various unigram and n-gram taggers, incorporating backoff, and train them on part of the Brown Corpus. 
a. Create three different combinations of the taggers. Test the accuracy of each combined tagger. Which combination works best? 
b. Try varying the size of the training corpus. How does it affect your results?

In [33]:
## Importing nltk and downloading brown corpus.
import nltk
from nltk.corpus import brown

In [34]:
nltk.download('brown')

[nltk_data] Downloading package brown to
[nltk_data]     C:\Users\utkar\AppData\Roaming\nltk_data...
[nltk_data]   Package brown is already up-to-date!


True

In [35]:
# Displaying the various categories present in the Brown Corpus
brown.categories()

['adventure',
 'belles_lettres',
 'editorial',
 'fiction',
 'government',
 'hobbies',
 'humor',
 'learned',
 'lore',
 'mystery',
 'news',
 'religion',
 'reviews',
 'romance',
 'science_fiction']

In [36]:
# Retrieve tagged sentences in the lore category
#getting the data ready
lore1 = brown.tagged_sents(categories='lore')
lore1

[[('In', 'IN'), ('American', 'JJ'), ('romance', 'NN'), (',', ','), ('almost', 'RB'), ('nothing', 'PN'), ('rates', 'VBZ'), ('higher', 'RBR'), ('than', 'CS'), ('what', 'WDT'), ('the', 'AT'), ('movie', 'NN'), ('men', 'NNS'), ('have', 'HV'), ('called', 'VBN'), ('``', '``'), ('meeting', 'VBG'), ('cute', 'JJ'), ("''", "''"), ('--', '--'), ('that', 'DT'), ('is', 'BEZ'), (',', ','), ('boy-meets-girl', 'NN'), ('seems', 'VBZ'), ('more', 'QL'), ('adorable', 'JJ'), ('if', 'CS'), ('it', 'PPS'), ("doesn't", 'DOZ*'), ('take', 'VB'), ('place', 'NN'), ('in', 'IN'), ('an', 'AT'), ('atmosphere', 'NN'), ('of', 'IN'), ('correct', 'JJ'), ('and', 'CC'), ('acute', 'JJ'), ('boredom', 'NN'), ('.', '.')], [('Just', 'QL'), ('about', 'RB'), ('the', 'AT'), ('most', 'QL'), ('enthralling', 'JJ'), ('real-life', 'NN'), ('example', 'NN'), ('of', 'IN'), ('meeting', 'VBG'), ('cute', 'JJ'), ('is', 'BEZ'), ('the', 'AT'), ('Charles', 'NP'), ('MacArthur-Helen', 'NP'), ('Hayes', 'NP'), ('saga', 'NN'), (':', ':'), ('reputedly',

In [37]:
# size of the selected part of the corpus
len(lore1)

4881

## creating a regular expression tagger:-

In [38]:
patterns = [
    (r'.*', 'NN'),                     # nouns
    (r'^-?[0-9]+(.[0-9]+)?$', 'CD'),   # cardinal numbers
    (r'(The|the|A|a|An|an)$', 'AT'),   # articles
    (r'.*able$', 'JJ'),                # adjectives
    (r'.*s$', 'NNS'),                  # plural nouns
    (r'.*\'s$', 'NN$'),                # possessive nouns
    (r'.*ing$', 'VBG'),                # gerunds
    (r'.*ed$', 'VBD'),                 # past tense verbs
    (r'.*es$', 'VBZ'),                 # 3rd singular present
    (r'.*ness$', 'NN'),                # nouns formed from adjectives
    (r'.*ly$', 'RB'),                  # adverbs
    (r'.*ould$', 'MD')                 # modals
]

In [39]:
# splitting into a train and test set (95% train, 5% test)
train = int(len(lore1) * 0.95)
train_sents = lore1[:train]
test_sents = lore1[train:]

In [40]:
# baseline accuracy (without backoff)
t0 = nltk.RegexpTagger(patterns)
t0.evaluate(test_sents)

0.12641631595985756

# Test various combination of taggers with backoff:

In [41]:
# regexp + unigram
t0 = nltk.RegexpTagger(patterns)
t1 = nltk.UnigramTagger(train_sents, backoff=t0)
t1.evaluate(test_sents)

0.8648429912593072

In [42]:
# bigram + unigram
t0 = nltk.BigramTagger(train_sents)
t1 = nltk.UnigramTagger(train_sents, backoff=t0)
t1.evaluate(test_sents)

0.5809323405632891

In [43]:
# regexp + unigram + bigram
t0 = nltk.RegexpTagger(patterns)
t1 = nltk.UnigramTagger(train_sents, backoff=t0)
t2 = nltk.BigramTagger(train_sents, backoff=t1)
t2.evaluate(test_sents)

0.8709938491421172

In [44]:
# regexp + unigram + bigram + trigram
t0 = nltk.RegexpTagger(patterns)
t1 = nltk.UnigramTagger(train_sents, backoff=t0)
t2 = nltk.BigramTagger(train_sents, backoff=t1)
t3 = nltk.TrigramTagger(train_sents, backoff=t2)
t3.evaluate(test_sents)

0.8727743606345095

In [46]:
# regexp + bigram + trigram
t0 = nltk.RegexpTagger(patterns)
t1 = nltk.BigramTagger(train_sents, backoff=t0)
t2 = nltk.TrigramTagger(train_sents, backoff=t1)
t2.evaluate(test_sents)

0.7593072191647783

In [45]:
# regexp + unigram + trigram
t0 = nltk.RegexpTagger(patterns)
t1 = nltk.UnigramTagger(train_sents, backoff=t0)
t2 = nltk.TrigramTagger(train_sents, backoff=t1)
t2.evaluate(test_sents)

0.8692133376497249

The highest accuracy is obtained on the combination with regexp + unigram + bigram + trigram which is 0.8727743606345095.<br>

The best combination has the simplest tagger (i.e, regexp) as the last backoff, and the most specific tagger (i.e, unigram/bigram/trigram) as the first to tag the sentences.<br>

The second highest is obtained on the combination with regexp + unigram + bigram which is 0.8709938491421172.<br>

# b. Try varying the size of the training corpus.


## Trying with a smaller training set:-

In [16]:
religion = brown.tagged_sents(categories='religion')
len(religion)

1716

In [17]:
# splitting into a train and test set (95% train, 5% test)
train = int(len(religion) * 0.95)
train_sents = religion[:train]
test_sents = religion[train:]

In [19]:
# second best accuracy
# regexp + unigram + bigram
t0 = nltk.RegexpTagger(patterns)
t1 = nltk.UnigramTagger(train_sents, backoff=t0)
t2 = nltk.BigramTagger(train_sents, backoff=t1)
t2.evaluate(test_sents)

0.8327053504144687

In [20]:
#best accuracy
# regexp + unigram + bigram + trigram
t0 = nltk.RegexpTagger(patterns)
t1 = nltk.UnigramTagger(train_sents, backoff=t0)
t2 = nltk.BigramTagger(train_sents, backoff=t1)
t3 = nltk.TrigramTagger(train_sents, backoff=t2)
t3.evaluate(test_sents)

0.8342125094197438

The accuracy is slightly lower on a smaller training set but the difference is not huge.

## Trying with a larger training set.<br>
(taking full Brown Corpus)

In [21]:
B_corpus = brown.tagged_sents()
len(B_corpus)

57340

In [22]:
# splitting into a train and test set (95% train, 5% test)
train = int(len(B_corpus) * 0.95)
train_sents = B_corpus[:train]
test_sents = B_corpus[train:]

In [23]:
#second best accuracy.
# regexp + unigram + bigram
t0 = nltk.RegexpTagger(patterns)
t1 = nltk.UnigramTagger(train_sents, backoff=t0)
t2 = nltk.BigramTagger(train_sents, backoff=t1)
t2.evaluate(test_sents)

0.9118542275243601

In [24]:
#best accuracy.
# regexp + unigram + bigram + trigram
t0 = nltk.RegexpTagger(patterns)
t1 = nltk.UnigramTagger(train_sents, backoff=t0)
t2 = nltk.BigramTagger(train_sents, backoff=t1)
t3 = nltk.TrigramTagger(train_sents, backoff=t2)
t3.evaluate(test_sents)

0.9123649214552735

## The accuracy increases when trained on much larger corpus.
<br>
------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------

##Clearly,we can observe that the accuracy is directly proportional to the size of corpus used for training.